File: mojo/android/system/src/org/chromium/mojo/system/impl/BaseRunLoop.java
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.mojo.system.impl;

import org.chromium.base.CalledByNative;
import org.chromium.base.JNINamespace;
import org.chromium.mojo.system.RunLoop;

import java.lang.Runnable;

/**
 * Implementation of {@link RunLoop} suitable for the base:: message loop implementation.
 */
@JNINamespace("mojo::android")
class BaseRunLoop implements RunLoop {
    /**
     * Pointer to the C run loop.
     */
    private long mRunLoopID;
    private final CoreImpl mCore;

    BaseRunLoop(CoreImpl core) {
        this.mCore = core;
        this.mRunLoopID = nativeCreateBaseRunLoop();
    }

    @Override
    public void run() {
        assert mRunLoopID != 0 : "The run loop cannot run once closed";
        nativeRun(mRunLoopID);
    }

    @Override
    public void runUntilIdle() {
        assert mRunLoopID != 0 : "The run loop cannot run once closed";
        nativeRunUntilIdle(mRunLoopID);
    }

    @Override
    public void quit() {
        assert mRunLoopID != 0 : "The run loop cannot be quitted run once closed";
        nativeQuit(mRunLoopID);
    }

    @Override
    public void postDelayedTask(Runnable runnable, long delay) {
        assert mRunLoopID != 0 : "The run loop cannot run tasks once closed";
        nativePostDelayedTask(mRunLoopID, runnable, delay);
    }

    @Override
    public void close() {
        if (mRunLoopID == 0) {
            return;
        }
        // We don't want to de-register a different run loop!
        assert mCore.getCurrentRunLoop() == this : "Only the current run loop can be closed";
        mCore.clearCurrentRunLoop();
        nativeDeleteMessageLoop(mRunLoopID);
        mRunLoopID = 0;
    }

    @CalledByNative
    private static void runRunnable(Runnable runnable) {
        runnable.run();
    }

    private native long nativeCreateBaseRunLoop();
    private native void nativeRun(long runLoopID);
    private native void nativeRunUntilIdle(long runLoopID);
    private native void nativeQuit(long runLoopID);
    private native void nativePostDelayedTask(long runLoopID, Runnable runnable, long delay);
    private native void nativeDeleteMessageLoop(long runLoopID);
}

File: mojo/android/system/base_run_loop.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/android/system/base_run_loop.h"

#include <jni.h>

#include "base/android/base_jni_registrar.h"
#include "base/android/jni_android.h"
#include "base/android/jni_registrar.h"
#include "base/bind.h"
#include "base/message_loop/message_loop.h"
#include "jni/BaseRunLoop_jni.h"
#include "mojo/common/message_pump_mojo.h"

namespace mojo {
namespace android {

namespace {

struct MessageLoopHolder {
  MessageLoopHolder() {
    if (base::MessageLoop::current()) {
      message_loop = base::MessageLoop::current();
      owned = false;
    } else {
      message_loop = new base::MessageLoop(common::MessagePumpMojo::Create());
      owned = true;
    }
  }

  ~MessageLoopHolder() {
    if (owned) {
      delete message_loop;
      message_loop = nullptr;
    }
  }

  base::MessageLoop* message_loop;
  bool owned;
};

}  // namespace

static jlong CreateBaseRunLoop(JNIEnv* env, jobject jcaller) {
  return reinterpret_cast<uintptr_t>(new MessageLoopHolder());
}

static void Run(JNIEnv* env, jobject jcaller, jlong runLoopID) {
  reinterpret_cast<MessageLoopHolder*>(runLoopID)->message_loop->Run();
}

static void RunUntilIdle(JNIEnv* env, jobject jcaller, jlong runLoopID) {
  reinterpret_cast<MessageLoopHolder*>(runLoopID)->message_loop->RunUntilIdle();
}

static void Quit(JNIEnv* env, jobject jcaller, jlong runLoopID) {
  reinterpret_cast<MessageLoopHolder*>(runLoopID)->message_loop->Quit();
}

static void RunJavaRunnable(
    const base::android::ScopedJavaGlobalRef<jobject>& runnable_ref) {
  Java_BaseRunLoop_runRunnable(base::android::AttachCurrentThread(),
                               runnable_ref.obj());
}

static void PostDelayedTask(JNIEnv* env,
                            jobject jcaller,
                            jlong runLoopID,
                            jobject runnable,
                            jlong delay) {
  base::android::ScopedJavaGlobalRef<jobject> runnable_ref;
  // ScopedJavaGlobalRef do not hold onto the env reference, so it is safe to
  // use it across threads. |RunJavaRunnable| will acquire a new JNIEnv before
  // running the Runnable.
  runnable_ref.Reset(env, runnable);
  reinterpret_cast<MessageLoopHolder*>(runLoopID)
      ->message_loop->PostDelayedTask(
          FROM_HERE, base::Bind(&RunJavaRunnable, runnable_ref),
          base::TimeDelta::FromMicroseconds(delay));
}

static void DeleteMessageLoop(JNIEnv* env, jobject jcaller, jlong runLoopID) {
  MessageLoopHolder* native_loop =
      reinterpret_cast<MessageLoopHolder*>(runLoopID);
  delete native_loop;
}

bool RegisterBaseRunLoop(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace android
}  // namespace mojo


